crates/pathfinder/src/storage/schema/revision_0011.rs

```diff
@@ -21,37 +21,136 @@ pub(crate) fn migrate(transaction: &Transaction) -> anyhow::Result<PostMigration
         "Decompressing transactions and fixing event addresses, this may take a while.",
     );
 
+    transaction
+        .execute("DROP INDEX starknet_events_from_address", [])
+        .context("Failed to drop the index before updates")?;
+
+    // the inner loop of the migration does a lot of updates based on transaction_hash and the
+    // index, so there should be an index for that, and it's unique since it can be.
+    transaction
+        .execute(
+            "CREATE UNIQUE INDEX temp_starknet_events_q ON starknet_events (idx, transaction_hash)",
+            [],
+        )
+        .context("create temporary index")?;
+
+    transaction
+        .execute("DROP TRIGGER starknet_events_au", [])
+        .context("Failed to drop after update trigger")?;
+
     let mut stmt = transaction
         .prepare("SELECT hash, receipt FROM starknet_transactions")
         .context("Prepare transaction query")?;
     let mut rows = stmt.query([])?;
 
+    let mut processed_rows = 0;
+    let start_of_run = std::time::Instant::now();
+    let mut start_of_batch = start_of_run;
+    let batch_size = if todo < 10000 { 0 } else { todo / 11 };
+    let mut decompression_time = std::time::Duration::default();
+    let mut parsing_time = std::time::Duration::default();
+
     while let Some(r) = rows.next()? {
         let transaction_hash = r.get_ref_unwrap("hash").as_blob()?;
         let receipt = r.get_ref_unwrap("receipt").as_blob()?;
+        let decompression_started = std::time::Instant::now();
         let receipt = zstd::decode_all(receipt).context("Decompress receipt")?;
-        let receipt: transaction::Receipt =
+        decompression_time += decompression_started.elapsed();
+
+        let parsing_started = std::time::Instant::now();
+        let receipt: LightReceipt =
             serde_json::de::from_slice(&receipt).context("Deserializing transaction receipt")?;
+        parsing_time += parsing_started.elapsed();
 
         receipt.events.into_iter().enumerate().try_for_each(
             |(idx, event)| -> anyhow::Result<_> {
-                transaction
-                    .execute(
-                        r"UPDATE starknet_events 
+                let mut stmt = transaction.prepare_cached(
```

Comment:
Can't we move this prepare() out of the loop?

Reply:
Good point, didn't even try that